Cython/Compiler/ModuleNode.py

```diff
@@ -810,18 +810,30 @@ def generate_module_preamble(self, env, options, cimported_modules, metadata, co
         self._put_setup_code(code, "PythonCompatibility")
         self._put_setup_code(code, "MathInitCode")
 
-        if options.c_line_in_traceback:
-            cinfo = "%s = %s; " % (Naming.clineno_cname, Naming.line_c_macro)
-        else:
-            cinfo = ""
-        code.putln("#define __PYX_MARK_ERR_POS(f_index, lineno) \\")
+        # Error handling and position macros.
         # Using "(void)cname" to prevent "unused" warnings.
-        code.putln("    { %s = %s[f_index]; (void)%s; %s = lineno; (void)%s; %s (void)%s; }" % (
-            Naming.filename_cname, Naming.filetable_cname, Naming.filename_cname,
-            Naming.lineno_cname, Naming.lineno_cname,
-            cinfo,
-            Naming.clineno_cname,
-        ))
+        mark_errpos_code = (
+            "#define __PYX_MARK_ERR_POS(f_index, lineno)  {"
+            f" {Naming.filename_cname} = {Naming.filetable_cname}[f_index];"
+            f" (void) {Naming.filename_cname};"
+            f" {Naming.lineno_cname} = lineno;"
+            f" (void) {Naming.lineno_cname};"
+            "%s"  # for C line info
+            f" (void) {Naming.clineno_cname}; "  # always suppress warnings
+            "}"
+        )
+        cline_info = f" {Naming.clineno_cname} = {Naming.line_c_macro};"
+
+        if not options.c_line_in_traceback:
+            code.putln("#ifndef CYTHON_CLINE_IN_TRACEBACK")
+            code.putln("#define CYTHON_CLINE_IN_TRACEBACK  0")
+            code.putln("#endif")
+        code.putln("#if !defined(CYTHON_CLINE_IN_TRACEBACK) || CYTHON_CLINE_IN_TRACEBACK")
```

Comment:
I found this a bit difficult to follow, so rewrote it in a way I found more obvious. I think the core behaviour is basically as you describe although some of the edge cases where the different options contradict each other might be different.

I've also added a fairly thorough set of tests I think (so feel free to undo my rewrite since the tests should at least confirm it's doing the right thing now)